imports/node-app/plugins/discount-codes/mutations/applyDiscountCodeToCart.js

```diff
@@ -0,0 +1,134 @@
+  const { Cart, Discounts } = collections;
+
+  // TODO: figure out the correct permission check here
+  // Should it be `discounts`, or `cart`?
+  // How do we determine this check if the user is the cart owner?
+  if (!userHasPermission(["admin", "owner", "discounts"], shopId)) {
+    throw new ReactionError("access-denied", "Access Denied");
+  }
+
+  let userCount = 0;
+  let orderCount = 0;
+  let cart = await getCart(context, shopId, cartId, { cartToken: token, throwIfNotFound: false });
+
+  // If we didn't find a cart, it means it belongs to another user,
+  // not the currently logged in user.
+  // Check to make sure current user has admin permission.
+  if (!cart) {
+    cart = await Cart.findOne({ _id: cartId });
+    if (!cart) {
+      throw new ReactionError("not-found", "Cart not found");
+    }
+
+    // TODO: figure out the correct permission check here
+    // Should it be `discounts`, or `cart`?
+    if (!userHasPermission(["owner", "admin", "discounts"], shopId)) {
+      throw new ReactionError("access-denied", "Access Denied");
+    }
+  }
+
+  const objectToApplyDiscount = cart;
+
+  // check to ensure discounts can only apply to single shop carts
+  // TODO: Remove this check after implementation of shop-by-shop discounts
+  // loop through all items and filter down to unique shops (in order to get participating shops in the order/cart)
+  const uniqueShopObj = objectToApplyDiscount.items.reduce((shopObj, item) => {
+    if (!shopObj[item.shopId]) {
+      shopObj[item.shopId] = true;
+    }
+    return shopObj;
+  }, {});
+  const participatingShops = Object.keys(uniqueShopObj);
+
+  if (participatingShops.length > 1) {
+    throw new ReactionError("not-implemented", "discounts.multiShopError", "Discounts cannot be applied to a multi-shop cart or order");
+  }
+
+  const discount = await Discounts.findOne({ code: discountCode });
+  if (!discount) throw new ReactionError("not-found", `No discount found for code ${discountCode}`);
+
+  const { conditions } = discount;
+  let accountLimitExceeded = false;
+  let discountLimitExceeded = false;
+
+  // existing usage count
+  if (discount.transactions) {
+    const users = Array.from(discount.transactions, (trans) => trans.userId);
+    const transactionCount = new Map([...new Set(users)].map((userX) => [userX, users.filter((userY) => userY === userX).length]));
+    const orders = Array.from(discount.transactions, (trans) => trans.cartId);
+    userCount = transactionCount.get(userId);
+    orderCount = orders.length;
+  }
+  // check limits
+  if (conditions) {
+    if (conditions.accountLimit) accountLimitExceeded = conditions.accountLimit <= userCount;
+    if (conditions.redemptionLimit) discountLimitExceeded = conditions.redemptionLimit <= orderCount;
+  }
+
+  // validate basic limit handling
+  if (accountLimitExceeded === true || discountLimitExceeded === true) {
+    return { i18nKeyLabel: "Code is expired", i18nKey: "discounts.codeIsExpired" };
```

Comment:
I'd prefer to throw a ReactionError and let the client translate the error to the i18n key.